homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,674 @@
+        await self.set_unavailable()
+
+    async def async_media_play(self) -> None:
+        """Begin playback."""
+        if self.client:
+            await self.client.play(self.player_id)
+
+    async def async_media_pause(self) -> None:
+        """Pause playback."""
+        if self.client:
+            await self.client.pause(self.player_id)
+
+    async def async_media_stop(self) -> None:
+        """Stop playback."""
+        if self.client:
+            await self.client.stop(self.player_id)
+
+    async def async_media_next_track(self) -> None:
+        """Skip to next track."""
+        if self.client:
+            await self.client.next(self.player_id)
+
+    async def async_media_previous_track(self) -> None:
+        """Skip to previous track."""
+        if self.client:
+            await self.client.previous(self.player_id)
+
+    async def async_media_seek(self, position: float) -> None:
+        """Send seek command."""
+        if self.client:
+            trackid = self._metadata.get("mpris:trackid")
+            if trackid:
+                await self.client.set_position(
+                    self.player_id,
+                    trackid,
+                    position,
+                )
+
+    async def update_state(
+        self,
+        new_state: str,
+    ):
+        """Update player state based on reports from the server."""
+        if new_state == self._attr_state:
+            return
+
+        _LOGGER.debug(
+            "Updating state from %s to %s",
+            self._attr_state,
+            new_state,
+        )
+        self._attr_state = new_state
+        if self.hass:
+            await self.async_update_ha_state(True)
+
+    async def update_metadata(self, new_metadata: dict[str, Any]):
+        """Update player metadata based on incoming metadata (a dict)."""
+        self._metadata = new_metadata
+        if "mpris:length" in self._metadata:
+            length: int | None = round(
+                float(self._metadata["mpris:length"]) / 1000 / 1000
+            )
+            if length is not None and length <= 0:
+                length = None
+        else:
+            length = None
+
+        self._attr_media_duration = length
+        self._attr_media_position = 0 if length is not None else None
+        self._attr_media_position_updated_at = dt_util.utcnow()
+
+        _LOGGER.debug("Setting media duration to %s", self._attr_media_duration)
+        _LOGGER.debug("Setting media position to %s", self._attr_media_position)
+
+        if self.hass:
+            await self.async_update_ha_state(True)
+
+    async def update_position(self, new_position: float):
+        """Update position."""
+        self._attr_media_position_updated_at = dt_util.utcnow()
+        self._attr_media_position = (
+            round(new_position) if new_position is not None else None
+        )
+        _LOGGER.debug("Setting media position to %s", self._attr_media_position)
+        if self.hass:
```

Comment:
We shouldn't write state before its added to hass

Reply:
I've seen this pattern everywhere in the codebase, so I just copied it.

What is the _correct_ way of doing this?  Thanks in advance for your kind help.

Reply:
I'm not sure what needs to be done here.  It's been a while since this comment was posted (over half a year).  All writes of state are conditioned to `self.hass` being available, which to my understanding only happens when the entity has been added to `hass` (there is no other way, to my knowledge, that the attribute `hass` can be not `None`).

If there is any change that still needs to be done, please let me know.

Reply:
It has been addressed.  I will be pushing the code now.